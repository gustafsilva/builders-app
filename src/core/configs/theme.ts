import { createTheme } from "@rneui/themed";

const theme = createTheme({
  lightColors: {},
  mode: "light",
});

export default theme;
