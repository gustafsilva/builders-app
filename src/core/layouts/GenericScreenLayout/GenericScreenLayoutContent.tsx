import React from "react";
import { useTheme } from "@rneui/themed";

import Box from "core/components/Box";
import Spinner from "core/components/Spinner";

export interface GenericScreenLayoutContentProps {
  loading?: boolean;
  children: React.ReactElement;
}

function GenericScreenLayoutContent({
  loading,
  children,
}: GenericScreenLayoutContentProps) {
  const { theme } = useTheme();

  if (loading) {
    return (
      <Box flex={1} justifyContent="center" alignItems="center">
        <Spinner color={theme.colors.secondary} />
      </Box>
    );
  }

  return children;
}

export default GenericScreenLayoutContent;
