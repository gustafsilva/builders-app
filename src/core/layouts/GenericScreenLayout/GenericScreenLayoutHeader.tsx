import React from "react";
import { useTheme } from "@rneui/themed";
import { useNavigation } from "@react-navigation/native";
import { TouchableOpacity } from "react-native-gesture-handler";
import { MaterialCommunityIcons } from "@expo/vector-icons";

import Box from "core/components/Box";

function GenericScreenLayoutHeader() {
  const { theme } = useTheme();
  const navigation = useNavigation();

  const goBack = () => navigation.goBack();

  return (
    <Box mt="view" ml="card">
      <TouchableOpacity onPress={goBack}>
        <MaterialCommunityIcons
          name="arrow-left"
          size={24}
          color={theme.colors.black}
        />
      </TouchableOpacity>
    </Box>
  );
}

export default GenericScreenLayoutHeader;
