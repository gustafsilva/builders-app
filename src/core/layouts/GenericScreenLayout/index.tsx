import React from "react";
import { StatusBar } from "expo-status-bar";

import SafeAreaContainer from "core/components/SafeAreaContainer";
import Container from "core/components/Container";

import GenericScreenLayoutHeader from "./GenericScreenLayoutHeader";
import GenericScreenLayoutContent, {
  GenericScreenLayoutContentProps,
} from "./GenericScreenLayoutContent";

export interface GenericScreenLayoutProps
  extends GenericScreenLayoutContentProps {
  headerShown?: boolean;
  statusBarStyle?: "light" | "dark";
  safeArea?: boolean;
}

function GenericScreenLayout({
  headerShown,
  statusBarStyle = "light",
  safeArea = true,
  ...props
}: GenericScreenLayoutProps) {
  const ExternalContainer = safeArea ? SafeAreaContainer : Container;

  return (
    <ExternalContainer>
      {headerShown && <GenericScreenLayoutHeader />}
      <GenericScreenLayoutContent {...props} />
      <StatusBar style={statusBarStyle} />
    </ExternalContainer>
  );
}

export default GenericScreenLayout;
