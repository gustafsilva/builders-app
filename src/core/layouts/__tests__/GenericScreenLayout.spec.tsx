import React from "react";
import { Text } from "react-native";
import { render, screen } from "@testing-library/react-native";
import { ReactTestInstance } from "react-test-renderer";
import { NavigationContainer } from "@react-navigation/native";
import { ThemeProvider } from "@rneui/themed";

import GenericScreenLayout from "core/layouts/GenericScreenLayout";
import theme from "core/configs/theme";

const TEST_ID_OF_LAYOUT = "GenericScreenLayout";
const TEXT = "Container Children";
const CHILDREN = <Text>{TEXT}</Text>;

describe("ü´Ä Core | üñºÔ∏è Layouts | *Ô∏è‚É£ <GenericScreenLayout>", function () {
  let genericScreenLayout: ReactTestInstance | null = null;

  describe("| without header", function () {
    beforeEach(function () {
      render(
        <GenericScreenLayout children={CHILDREN} testID={TEST_ID_OF_LAYOUT} />
      );

      genericScreenLayout = screen.getByTestId(TEST_ID_OF_LAYOUT);
    });

    it("should render successfully", () => {
      expect(genericScreenLayout).toBeDefined();
    });

    it("should render correctly", () => {
      const screenAsJson = screen.toJSON();
      expect(screenAsJson).toMatchSnapshot();
    });

    it("should render with children", () => {
      expect(screen.getByText(TEXT)).toBeTruthy();
    });
  });

  describe("| with header", function () {
    beforeEach(function () {
      render(
        <ThemeProvider theme={theme}>
          <NavigationContainer>
            <GenericScreenLayout
              children={CHILDREN}
              testID={TEST_ID_OF_LAYOUT}
              headerShown
            />
          </NavigationContainer>
        </ThemeProvider>
      );

      genericScreenLayout = screen.getByTestId(TEST_ID_OF_LAYOUT);
    });

    it("should render successfully", () => {
      expect(genericScreenLayout).toBeDefined();
    });

    it("should render correctly", () => {
      const screenAsJson = screen.toJSON();
      expect(screenAsJson).toMatchSnapshot();
    });

    it("should render with children", () => {
      expect(screen.getByText(TEXT)).toBeTruthy();
    });
  });
});
