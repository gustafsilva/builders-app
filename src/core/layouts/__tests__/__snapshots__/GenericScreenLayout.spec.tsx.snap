// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`🫀 Core | 🖼️ Layouts | *️⃣ <GenericScreenLayout> | with header should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "flex": 1,
    }
  }
  testID="GenericScreenLayout"
>
  <View
    ml="card"
    mt="view"
    style={
      Object {
        "alignItems": undefined,
        "backgroundColor": undefined,
        "borderRadius": undefined,
        "borderWidth": undefined,
        "flex": undefined,
        "flexDirection": undefined,
        "height": undefined,
        "justifyContent": undefined,
        "margin": undefined,
        "marginBottom": undefined,
        "marginHorizontal": undefined,
        "marginLeft": 16,
        "marginRight": undefined,
        "marginTop": 24,
        "marginVertical": undefined,
        "opacity": undefined,
        "padding": undefined,
        "paddingBottom": undefined,
        "paddingHorizontal": undefined,
        "paddingLeft": undefined,
        "paddingRight": undefined,
        "paddingTop": undefined,
        "paddingVertical": undefined,
        "width": undefined,
      }
    }
  >
    <RNGestureHandlerButton
      collapsable={false}
      delayLongPress={600}
      enabled={true}
      exclusive={true}
      handlerTag={2}
      handlerType="NativeViewGestureHandler"
      onGestureEvent={[Function]}
      onGestureHandlerEvent={[Function]}
      onGestureHandlerStateChange={[Function]}
      onHandlerStateChange={[Function]}
      rippleColor={0}
      touchSoundDisabled={false}
    >
      <View
        accessible={true}
        collapsable={false}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <Text
          allowFontScaling={false}
          selectable={false}
          style={
            Array [
              Object {
                "color": "#FFFFFF",
                "fontSize": 24,
              },
              undefined,
              Object {
                "fontFamily": "material-community",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
              Object {},
            ]
          }
        >
          󰁍
        </Text>
      </View>
    </RNGestureHandlerButton>
  </View>
  <Text>
    Container Children
  </Text>
</RNCSafeAreaView>
`;

exports[`🫀 Core | 🖼️ Layouts | *️⃣ <GenericScreenLayout> | without header should render correctly 1`] = `
<RNCSafeAreaView
  edges={
    Array [
      "bottom",
      "left",
      "right",
      "top",
    ]
  }
  style={
    Object {
      "flex": 1,
    }
  }
  testID="GenericScreenLayout"
>
  <Text>
    Container Children
  </Text>
</RNCSafeAreaView>
`;
